<a href="https://colab.research.google.com/github/Shivkumar-777/Edu-Python/blob/main/Books.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
class EnergySystem:

  def __init__(self, building_name, energy_consumption, emission_factor):
    self.building_name = building_name
    self.energy_consumption = energy_consumption
    self.emission_factor = emission_factor

building1 = EnergySystem("Building A", 5000, 0.45)
building2 = EnergySystem("Building B", 8000, 0.60)
building3 = EnergySystem("Building C", 3000, 0.30)

total_energy_consumption = building1.energy_consumption + building2.energy_consumption + building3.energy_consumption
total_emission = building1.energy_consumption * building1.emission_factor + building2.energy_consumption * building2.emission_factor + building3.energy_consumption * building3.emission_factor

print("Total Energy Consumption:", total_energy_consumption, "kWh")
print("Total Emission:", total_emission, "kg CO2e")

def calculate_carbon_footprint(self):
  return self.energy_consumption * self.emission_factor

def calculate_energy_savings(self):
  return self.energy_consumption * 0.10

  print("Carbon Footprint:", calculate_carbon_footprint(building1), "kg CO2e")
  print("Energy Savings:", calculate_energy_savings(building1), "kWh")

Total Energy Consumption: 16000 kWh
Total Emission: 7950.0 kg CO2e


In [ ]:
# Inheritance

class SolarEnergySystem(EnergySystem):
  def __init__(self, building_name, energy_consumption, emission_factor, solar_production):
    super().__init__(building_name, energy_consumption, emission_factor)
    self.solar_production = solar_production

  def net_energy_consumption(self):
    return self.energy_consumption - self.solar_production

solar_building1 = SolarEnergySystem("Solar Building A", 5000, 0.45, 1500)
net_consumption = solar_building1.net_energy_consumption()

print("Net Energy Consumption:", net_consumption, "kWh")

Net Energy Consumption: 3500 kWh


In [ ]:
# Encapsulation

class EnergySystem:
  def __init__(self, building_name, energy_consumption, emission_factor):
    self._building_name = building_name
    self._energy_consumption = energy_consumption
    self._emission_factor = emission_factor

  def get_energy_consumption(self):
    return self._energy_consumption

  def calculate_carbon_footprint(self):
    return self._energy_consumption * self._emission_factor

  building = EnergySystem("Building A", 5000, 0.45)
  print("Building Name:", building._building_name)
  print("Energy Consumption:", building.get_energy_consumption(), "kWh")

Building Name: Building A
Energy Consumption: 5000 kWh


In [ ]:
# Polymorphism

class SolarEnergySystem(EnergySystem):
  def __init__(self, building_name, energy_consumption, emission_factor, solar_production):
    super().__init__(building_name, energy_consumption, emission_factor)
    self.solar_production = solar_production

  def calculate_carbon_footprint(self):
    net_consumption = self._energy_consumption - self.solar_production
    return net_consumption * self._emission_factor

solar_building1 = SolarEnergySystem("Solar Building A", 5000, 0.45, 1500)
print("Carbon Footprint:", solar_building1.calculate_carbon_footprint(), "kg CO2e")

Carbon Footprint: 1575.0 kg CO2e


### Smart Energy Consumption Analyzer

In [17]:
# 1. Stores energy consumption using appropriate variables and data types
# Let's represent each building's data as a dictionary in a list

buildings_data = [
    {"name": "Building A", "consumption": 2500},
    {"name": "Building B", "consumption": 1500},
    {"name": "Building C", "consumption": 4000},
    {"name": "Building D", "consumption": 800},
    {"name": "Building E", "consumption": 3000},
    {"name": "Building F", "consumption": 5500}
]

critical_grid_limit = 10000  # kWh

classified_buildings = []
total_consumption_so_far = 0


In [18]:
# 2. Uses a loop to process energy data for multiple buildings
# 3. Uses if–elif–else to classify buildings
# 5. Uses break if total consumption exceeds a critical grid limit

print("--- Processing Energy Data ---")
for building in buildings_data:
    building_name = building["name"]
    consumption = building["consumption"]

    # Classify building based on consumption
    if consumption < 1000:
        classification = "Energy Efficient"
    elif 1000 <= consumption <= 3000:
        classification = "Moderate Consumption"
    else:
        classification = "Energy Intensive"

    classified_buildings.append({
        "name": building_name,
        "consumption": consumption,
        "classification": classification
    })

    total_consumption_so_far += consumption

    print(f"  {building_name}: {consumption} kWh - {classification}")

    if total_consumption_so_far > critical_grid_limit:
        print(f"  [WARNING] Critical grid limit ({critical_grid_limit} kWh) exceeded! Stopping processing.")
        break

print(f"Total consumption processed so far: {total_consumption_so_far} kWh")
print("------------------------------")


--- Processing Energy Data ---
  Building A: 2500 kWh - Moderate Consumption
  Building B: 1500 kWh - Moderate Consumption
  Building C: 4000 kWh - Energy Intensive
  Building D: 800 kWh - Energy Efficient
  Building E: 3000 kWh - Moderate Consumption
  [WARNING] Critical grid limit (10000 kWh) exceeded! Stopping processing.
Total consumption processed so far: 11800 kWh
------------------------------


In [19]:
# 4. Uses a function to compute total and average consumption
def calculate_consumption_stats(buildings_list):
    total_cons = sum(b["consumption"] for b in buildings_list)
    average_cons = total_cons / len(buildings_list) if buildings_list else 0
    return total_cons, average_cons


final_total_consumption, final_average_consumption = calculate_consumption_stats(classified_buildings)


In [20]:
# 6. Prints a sustainability report

print("\n--- Sustainability Report ---")
print(f"Critical Grid Limit: {critical_grid_limit} kWh\n")

print("Building-wise Classification:")
for building in classified_buildings:
    print(f"  - {building['name']}: {building['consumption']} kWh ({building['classification']})")

print(f"\nOverall Total Consumption: {final_total_consumption} kWh")
print(f"Overall Average Consumption per Building: {final_average_consumption:.2f} kWh")

if total_consumption_so_far > critical_grid_limit:
    print("\nAction Required: The critical grid limit was exceeded. Immediate measures for energy reduction or grid management are recommended.")
else:
    print("\nStatus: Overall energy consumption is within the critical grid limit.")

print("-------------------------------")



--- Sustainability Report ---
Critical Grid Limit: 10000 kWh

Building-wise Classification:
  - Building A: 2500 kWh (Moderate Consumption)
  - Building B: 1500 kWh (Moderate Consumption)
  - Building C: 4000 kWh (Energy Intensive)
  - Building D: 800 kWh (Energy Efficient)
  - Building E: 3000 kWh (Moderate Consumption)

Overall Total Consumption: 11800 kWh
Overall Average Consumption per Building: 2360.00 kWh

Action Required: The critical grid limit was exceeded. Immediate measures for energy reduction or grid management are recommended.
-------------------------------


# Task
Build a Renewable Energy Decision Engine in Python, demonstrating OOP principles. This will involve defining an `EnergySource` base class, deriving `SolarEnergy`, `WindEnergy`, and `HydroEnergy` subclasses, implementing polymorphic `calculate_availability` methods, and developing an energy allocation logic that prioritizes sources to meet demand. Finally, simulate and report on the energy allocation.

## Define EnergySource Base Class

### Subtask:
Create the `EnergySource` base class with a constructor to initialize common attributes like source name and a method for power output, demonstrating encapsulation.


**Reasoning**:
The subtask requires defining an `EnergySource` base class with a constructor to initialize protected attributes for source name and current output, and a public method to retrieve the current output. This code block directly implements these requirements.



In [21]:
class EnergySource:
  def __init__(self, name, current_output):
    self._name = name  # Protected attribute for source name
    self._current_output = current_output  # Protected attribute for current power output

  def get_current_output(self):
    """Returns the current power output of the energy source."""
    return self._current_output

# Example usage:
solar_panel = EnergySource("Solar Panel Array", 1500) # 1500 kW
wind_turbine = EnergySource("Wind Turbine Farm", 3000) # 3000 kW

print(f"{solar_panel._name} current output: {solar_panel.get_current_output()} kW")
print(f"{wind_turbine._name} current output: {wind_turbine.get_current_output()} kW")

Solar Panel Array current output: 1500 kW
Wind Turbine Farm current output: 3000 kW


## Derive Subclasses (Solar, Wind, Hydro)

### Subtask:
Implement `SolarEnergy`, `WindEnergy`, and `HydroEnergy` subclasses, inheriting from `EnergySource`. Each subclass will have its own constructor and potentially unique attributes. This will demonstrate inheritance.


**Reasoning**:
The subtask requires defining three subclasses (`SolarEnergy`, `WindEnergy`, `HydroEnergy`) that inherit from `EnergySource`, each with its own constructor and specific attributes, and then demonstrating their instantiation. This code block directly implements these requirements.



In [22]:
class SolarEnergy(EnergySource):
  def __init__(self, name, current_output, panel_efficiency):
    super().__init__(name, current_output)
    self._panel_efficiency = panel_efficiency # Unique attribute for SolarEnergy

class WindEnergy(EnergySource):
  def __init__(self, name, current_output, wind_speed):
    super().__init__(name, current_output)
    self._wind_speed = wind_speed # Unique attribute for WindEnergy

class HydroEnergy(EnergySource):
  def __init__(self, name, current_output, water_flow_rate):
    super().__init__(name, current_output)
    self._water_flow_rate = water_flow_rate # Unique attribute for HydroEnergy

# Example usage to demonstrate inheritance and unique attributes
solar_farm = SolarEnergy("Desert Solar Farm", 5000, 0.22) # 5000 kW, 22% efficiency
wind_park = WindEnergy("Offshore Wind Park", 7500, 15) # 7500 kW, 15 m/s wind speed
hydro_plant = HydroEnergy("River Hydro Plant", 10000, 500) # 10000 kW, 500 m3/s water flow

print(f"\n{solar_farm._name}: Current Output = {solar_farm.get_current_output()} kW, Panel Efficiency = {solar_farm._panel_efficiency * 100}%")
print(f"{wind_park._name}: Current Output = {wind_park.get_current_output()} kW, Wind Speed = {wind_park._wind_speed} m/s")
print(f"{hydro_plant._name}: Current Output = {hydro_plant.get_current_output()} kW, Water Flow Rate = {hydro_plant._water_flow_rate} m3/s")


Desert Solar Farm: Current Output = 5000 kW, Panel Efficiency = 22.0%
Offshore Wind Park: Current Output = 7500 kW, Wind Speed = 15 m/s
River Hydro Plant: Current Output = 10000 kW, Water Flow Rate = 500 m3/s


## Implement Polymorphic Energy Availability

### Subtask:
Define a polymorphic method `calculate_availability` in each subclass. This method will return the current available power, potentially with different logic for each energy source (e.g., solar dependent on sunlight, wind on wind speed, hydro on water flow). This will demonstrate polymorphism.


**Reasoning**:
The subtask requires defining a polymorphic `calculate_availability` method in the base class and overriding it in the subclasses with specific logic, then demonstrating its behavior. This code block directly implements all these requirements by modifying the existing classes and adding new examples to showcase the polymorphism.



In [23]:
class EnergySource:
  def __init__(self, name, current_output):
    self._name = name  # Protected attribute for source name
    self._current_output = current_output  # Protected attribute for current power output

  def get_current_output(self):
    """Returns the current power output of the energy source."""
    return self._current_output

  def calculate_availability(self):
    """Abstract method to be overridden by subclasses to calculate available power."""
    raise NotImplementedError("Subclasses must implement calculate_availability method")

class SolarEnergy(EnergySource):
  def __init__(self, name, current_output, panel_efficiency):
    super().__init__(name, current_output)
    self._panel_efficiency = panel_efficiency # Unique attribute for SolarEnergy

  def calculate_availability(self):
    if self._panel_efficiency > 0.20:
      return self._current_output
    else:
      return self._current_output * 0.50

class WindEnergy(EnergySource):
  def __init__(self, name, current_output, wind_speed):
    super().__init__(name, current_output)
    self._wind_speed = wind_speed # Unique attribute for WindEnergy

  def calculate_availability(self):
    if self._wind_speed >= 10:
      return self._current_output
    else:
      return self._current_output * 0.70

class HydroEnergy(EnergySource):
  def __init__(self, name, current_output, water_flow_rate):
    super().__init__(name, current_output)
    self._water_flow_rate = water_flow_rate # Unique attribute for HydroEnergy

  def calculate_availability(self):
    if self._water_flow_rate >= 400:
      return self._current_output
    else:
      return self._current_output * 0.90

# --- Demonstrating Polymorphism ---
print("\n--- Demonstrating Polymorphic Availability ---")

# Solar Energy examples
solar_high_efficiency = SolarEnergy("High-Efficiency Solar Farm", 6000, 0.25) # Efficiency > 0.20
solar_low_efficiency = SolarEnergy("Low-Efficiency Solar Farm", 4000, 0.18)   # Efficiency <= 0.20

print(f"{solar_high_efficiency._name}: Current Output = {solar_high_efficiency.get_current_output()} kW, Efficiency = {solar_high_efficiency._panel_efficiency*100}%, Available Power = {solar_high_efficiency.calculate_availability():.2f} kW")
print(f"{solar_low_efficiency._name}: Current Output = {solar_low_efficiency.get_current_output()} kW, Efficiency = {solar_low_efficiency._panel_efficiency*100}%, Available Power = {solar_low_efficiency.calculate_availability():.2f} kW")

# Wind Energy examples
wind_high_speed = WindEnergy("High-Wind Park", 8000, 12) # Wind speed >= 10 m/s
wind_low_speed = WindEnergy("Low-Wind Park", 5000, 8)   # Wind speed < 10 m/s

print(f"{wind_high_speed._name}: Current Output = {wind_high_speed.get_current_output()} kW, Wind Speed = {wind_high_speed._wind_speed} m/s, Available Power = {wind_high_speed.calculate_availability():.2f} kW")
print(f"{wind_low_speed._name}: Current Output = {wind_low_speed.get_current_output()} kW, Wind Speed = {wind_low_speed._wind_speed} m/s, Available Power = {wind_low_speed.calculate_availability():.2f} kW")

# Hydro Energy examples
hydro_high_flow = HydroEnergy("High-Flow Hydro Plant", 12000, 450) # Water flow >= 400 m3/s
hydro_low_flow = HydroEnergy("Low-Flow Hydro Plant", 9000, 350)   # Water flow < 400 m3/s

print(f"{hydro_high_flow._name}: Current Output = {hydro_high_flow.get_current_output()} kW, Water Flow Rate = {hydro_high_flow._water_flow_rate} m3/s, Available Power = {hydro_high_flow.calculate_availability():.2f} kW")
print(f"{hydro_low_flow._name}: Current Output = {hydro_low_flow.get_current_output()} kW, Water Flow Rate = {hydro_low_flow._water_flow_rate} m3/s, Available Power = {hydro_low_flow.calculate_availability():.2f} kW")



--- Demonstrating Polymorphic Availability ---
High-Efficiency Solar Farm: Current Output = 6000 kW, Efficiency = 25.0%, Available Power = 6000.00 kW
Low-Efficiency Solar Farm: Current Output = 4000 kW, Efficiency = 18.0%, Available Power = 2000.00 kW
High-Wind Park: Current Output = 8000 kW, Wind Speed = 12 m/s, Available Power = 8000.00 kW
Low-Wind Park: Current Output = 5000 kW, Wind Speed = 8 m/s, Available Power = 3500.00 kW
High-Flow Hydro Plant: Current Output = 12000 kW, Water Flow Rate = 450 m3/s, Available Power = 12000.00 kW
Low-Flow Hydro Plant: Current Output = 9000 kW, Water Flow Rate = 350 m3/s, Available Power = 8100.00 kW


## Implement Energy Allocation Logic

### Subtask:
Develop a function or class method that takes a list of energy sources and a demand. It will use conditional logic to prioritize energy sources (e.g., prioritize hydro, then solar, then wind) and allocate power using loops until the demand is met or all sources are depleted.


**Reasoning**:
The subtask requires implementing the `allocate_energy` function as described in the instructions, which includes categorizing energy sources, prioritizing them, and allocating power until demand is met or sources are depleted. This code block defines the function and provides example usage to demonstrate its functionality.



In [24]:
def allocate_energy(energy_sources, demand):
  allocated_power = {}
  unmet_demand = demand

  hydro_sources = []
  solar_sources = []
  wind_sources = []

  # Categorize energy sources
  for source in energy_sources:
    if isinstance(source, HydroEnergy):
      hydro_sources.append(source)
    elif isinstance(source, SolarEnergy):
      solar_sources.append(source)
    elif isinstance(source, WindEnergy):
      wind_sources.append(source)

  # Define prioritization order
  prioritized_categories = [hydro_sources, solar_sources, wind_sources]

  # Allocate power based on prioritization
  for category in prioritized_categories:
    if unmet_demand <= 0:
      break # Demand met
    for source in category:
      if unmet_demand <= 0:
        break # Demand met

      available_power = source.calculate_availability()
      power_to_allocate = min(unmet_demand, available_power)

      allocated_power[source._name] = power_to_allocate
      unmet_demand -= power_to_allocate

  return allocated_power, unmet_demand

# --- Example Usage ---
print("\n--- Demonstrating Energy Allocation Logic ---")

# Instantiate some energy sources
sources_for_allocation = [
    SolarEnergy("Small Solar Farm", 1000, 0.20),       # 1000 * 0.5 = 500 kW available
    WindEnergy("Coastal Wind Farm", 2000, 15),       # 2000 kW available
    HydroEnergy("Mountain Hydro Plant", 3000, 300),   # 3000 * 0.9 = 2700 kW available
    SolarEnergy("Large Solar Farm", 4000, 0.25),       # 4000 kW available
    WindEnergy("Inland Wind Farm", 1500, 8),         # 1500 * 0.7 = 1050 kW available
    HydroEnergy("Valley Hydro Plant", 2500, 450)    # 2500 kW available
]

demand1 = 7000 # kW
print(f"\nAttempting to meet demand of {demand1} kW")
allocated1, unmet1 = allocate_energy(sources_for_allocation, demand1)
print("Allocated Power:", allocated1)
print(f"Unmet Demand: {unmet1} kW")

demand2 = 3000 # kW
print(f"\nAttempting to meet demand of {demand2} kW")
allocated2, unmet2 = allocate_energy(sources_for_allocation, demand2)
print("Allocated Power:", allocated2)
print(f"Unmet Demand: {unmet2} kW")

demand3 = 15000 # kW
print(f"\nAttempting to meet demand of {demand3} kW")
allocated3, unmet3 = allocate_energy(sources_for_allocation, demand3)
print("Allocated Power:", allocated3)
print(f"Unmet Demand: {unmet3} kW")



--- Demonstrating Energy Allocation Logic ---

Attempting to meet demand of 7000 kW
Allocated Power: {'Mountain Hydro Plant': 2700.0, 'Valley Hydro Plant': 2500, 'Small Solar Farm': 500.0, 'Large Solar Farm': 1300.0}
Unmet Demand: 0.0 kW

Attempting to meet demand of 3000 kW
Allocated Power: {'Mountain Hydro Plant': 2700.0, 'Valley Hydro Plant': 300.0}
Unmet Demand: 0.0 kW

Attempting to meet demand of 15000 kW
Allocated Power: {'Mountain Hydro Plant': 2700.0, 'Valley Hydro Plant': 2500, 'Small Solar Farm': 500.0, 'Large Solar Farm': 4000, 'Coastal Wind Farm': 2000, 'Inland Wind Farm': 1050.0}
Unmet Demand: 2250.0 kW


## Simulate and Report

### Subtask:
Simulate the energy allocation for a given demand, printing which sources are used and how much power is drawn from each, along with any unmet demand.


**Reasoning**:
The subtask requires defining a new function `generate_allocation_report` to present the results of the energy allocation simulation, and then calling it with the specified variables. This code block will define and execute the function.



In [25]:
def generate_allocation_report(original_demand, allocated_power, unmet_demand):
  print(f"\n--- Energy Allocation Report for Demand: {original_demand} kW ---")

  if not allocated_power:
      print("No power was allocated from any source.")
  else:
      print("Power Allocated From:")
      for source_name, power in allocated_power.items():
          print(f"  - {source_name}: {power:.2f} kW")

  total_allocated = sum(allocated_power.values())
  print(f"\nTotal Power Allocated: {total_allocated:.2f} kW")
  print(f"Unmet Demand: {unmet_demand:.2f} kW")

  if unmet_demand <= 0:
      print("Conclusion: The energy demand was successfully met.")
  else:
      print("Conclusion: The energy demand was not fully met, and additional resources are required.")
  print("---------------------------------------------------")

# Call the function for demand3, allocated3, and unmet3
generate_allocation_report(demand3, allocated3, unmet3)


--- Energy Allocation Report for Demand: 15000 kW ---
Power Allocated From:
  - Mountain Hydro Plant: 2700.00 kW
  - Valley Hydro Plant: 2500.00 kW
  - Small Solar Farm: 500.00 kW
  - Large Solar Farm: 4000.00 kW
  - Coastal Wind Farm: 2000.00 kW
  - Inland Wind Farm: 1050.00 kW

Total Power Allocated: 12750.00 kW
Unmet Demand: 2250.00 kW
Conclusion: The energy demand was not fully met, and additional resources are required.
---------------------------------------------------


## Final Task

### Subtask:
Summarize the implemented Renewable Energy Decision Engine, highlighting the use of OOP principles and its functionality.


## Summary:

### Q&A
The implemented Renewable Energy Decision Engine is a Python-based system designed to simulate the allocation of power from various renewable energy sources to meet a specific energy demand. It effectively demonstrates key Object-Oriented Programming (OOP) principles:
*   **Encapsulation:** The `EnergySource` base class encapsulates common attributes like `_name` and `_current_output` and provides a public method `get_current_output()` for controlled access.
*   **Inheritance:** `SolarEnergy`, `WindEnergy`, and `HydroEnergy` are specialized subclasses that inherit from the `EnergySource` base class, allowing them to reuse common functionalities while introducing their unique attributes (e.g., `_panel_efficiency` for solar, `_wind_speed` for wind).
*   **Polymorphism:** Each subclass overrides the `calculate_availability()` method from the base class, implementing its own distinct logic to determine available power based on its specific environmental factors (e.g., solar efficiency, wind speed, water flow rate). The base class ensures this method is implemented by subclasses using `NotImplementedError`.

The engine's functionality includes:
1.  **Modeling Energy Sources:** Defining different renewable energy types with their specific characteristics.
2.  **Calculating Availability:** Dynamically determining the available power from each source based on simplified conditions.
3.  **Prioritized Allocation:** Distributing available power to meet demand following a predefined priority order (Hydro -> Solar -> Wind).
4.  **Reporting:** Generating a clear report detailing how much power was allocated from each source, the total allocated, and any remaining unmet demand.

### Data Analysis Key Findings
*   The `EnergySource` base class was successfully created, encapsulating `_name` and `_current_output` attributes, and providing a `get_current_output()` method.
*   `SolarEnergy`, `WindEnergy`, and `HydroEnergy` subclasses were correctly derived, each adding unique attributes like `_panel_efficiency`, `_wind_speed`, and `_water_flow_rate` respectively, demonstrating inheritance.
*   Polymorphic `calculate_availability()` methods were implemented in each subclass:
    *   `SolarEnergy`: Returns full output if `_panel_efficiency` > 0.20, otherwise 50% of output.
    *   `WindEnergy`: Returns full output if `_wind_speed` \>= 10 m/s, otherwise 70% of output.
    *   `HydroEnergy`: Returns full output if `_water_flow_rate` \>= 400 m³/s, otherwise 90% of output.
*   The energy allocation logic prioritizes sources in the order of Hydro, then Solar, then Wind.
*   Simulation for a 7000 kW demand resulted in the demand being fully met (0 kW unmet).
*   Simulation for a 3000 kW demand resulted in the demand being fully met (0 kW unmet).
*   Simulation for a 15000 kW demand partially met the demand, allocating 12750 kW from all available sources and leaving 2250 kW unmet.
*   A comprehensive report function was developed to summarize allocated power per source, total allocated power, and unmet demand.

### Insights or Next Steps
*   Enhance the `calculate_availability` logic to incorporate more dynamic and realistic environmental data (e.g., real-time weather, historical patterns) rather than simple thresholds.
*   Introduce configurable prioritization for energy sources, allowing users to define different allocation strategies based on factors like cost, environmental impact, or operational constraints.
